### ISSUE TITLE
spotify use chromium for videos ad

### ISSUE THREAD

**Issue by 021hossein** [danigm/spotify-ad-blocker#17]
Spotify use chromium for videos ad, but the extension mutes  just Spotify.

![Screenshot from 2021-12-16 21-39-45](https://user-images.githubusercontent.com/24650329/146426825-bdd421a5-51f2-421c-a6ee-3385a2782ff0.png)

**Comment by bmaupin** [comment]
I just saw a video ad so I dumped the MPRIS data. Looks like the filter for `mpris:trackid` should still work, so it would just be the mixer sink filter that would need to be modified. @danigm Does it seem feasible to update the mixer sink filter to include `chromium` as well? I'm trying to think of what the unintended effects would be ...

Here's the MPRIS data dump:

```
$ dbus-send --print-reply --dest=org.mpris.MediaPlayer2.spotify /org/mpris/MediaPlayer2 org.freedesktop.DBus.Properties.Get  string:'org.mpris.MediaPlayer2.Player' string:'Metadata'
method return time=1686840708.549485 sender=:1.438 -> destination=:1.444 serial=67 reply_serial=2
   variant       array [
         dict entry(
            string "mpris:trackid"
            variant                string "/com/spotify/ad/252fb46a7c740baa771dc45139abe654"
         )
         dict entry(
            string "mpris:length"
            variant                uint64 30054000
         )
         dict entry(
            string "mpris:artUrl"
            variant                string ""
         )
         dict entry(
            string "xesam:album"
            variant                string ""
         )
         dict entry(
            string "xesam:albumArtist"
            variant                array [
                  string ""
               ]
         )
         dict entry(
            string "xesam:artist"
            variant                array [
                  string ""
               ]
         )
         dict entry(
            string "xesam:autoRating"
            variant                double 0
         )
         dict entry(
            string "xesam:discNumber"
            variant                int32 0
         )
         dict entry(
            string "xesam:title"
            variant                string "McDonald's"
         )
         dict entry(
            string "xesam:trackNumber"
            variant                int32 0
         )
         dict entry(
            string "xesam:url"
            variant                string "https://open.spotify.com/ad/252fb46a7c740baa771dc45139abe654"
         )
      ]
```

**Comment by danigm** [comment]
I don't know if it's possible to detect which chromium is the one related to spotify... In other case, muting all chromiums could be a solution, I think it's not a common user case to have music and something else in chromium browser, but to solve that user case maybe we can add a configuration to enable/disable the chromium mute.

**Comment by bmaupin** [comment]
The cleanest way to do this would be to see if we could somehow associate a particular stream with its application, then we would know which chromium streams belong to Spotify.

If you run `pactl list sink-inputs` it has some interesting properties that seem to link the stream with its application, for example:

```
                application.name = "spotify"
                application.process.binary = "spotify"
```

Unfortunately these properties don't seem to be available in the Gnome stream object. It does have `get_application_id()` but it always seems to be null. I also found `get_description()` but for the chromium streams it just returns `Playback`.

Maybe this information is available via another Gnome API? I haven't been able to find anything but I'm not very familiar with the Gnome APIs.

**Comment by bmaupin** [comment]
It took me a few days to finally get a video ad, but this is what I got:

```
name=Chromium
application_id=null
description=Playback
get_icon_name=chromium-browser
```

compared to a spotify audio stream:

```
name=spotify
application_id=null
description=Spotify
get_icon_name=audio
```

**Comment by danigm** [comment]
So, there's no way to block video ads, because everything identifies as Chromium so we can't differentiate from an actual browser playing a video.

**Comment by bmaupin** [comment]
While writing a response I found something! Apparently the information I found through pactl is also available through wpctl, which I believe uses the Wireplumber API for Pipewire instead of the Pulseaudio API:

```
$ wpctl status | grep -A 10 Streams | grep -i spotify
        79. spotify

$ wpctl inspect 79 | grep -i spotify
  * application.name = "spotify"
    application.process.binary = "spotify"
    media.name = "Spotify"
  * node.name = "spotify"
```

Better yet, Wireplumber seems to have Gjs bindings: https://gjs-docs.gnome.org/wp05~0.5/

I have no idea how to use the API but I'll play with it and see if I can get any useful information out of it.

**Comment by bmaupin** [comment]
I don't get many video ads but I finally got one and grabbed the data from pipewire. It does seem that `application.process.binary` could be used to link a video ad to spotify:

```
wpctl status | grep -A 10 Streams | grep -i chromium
        78. Chromium
```

```
wpctl inspect 78 | grep spotify
    application.process.binary = "spotify"
```

```
   application.name = "Chromium"
   media.name = "Playback"
   node.name = "Chromium"
```

> Using subprocess will change a bit the plugin, but it's another option to simplify and do not depend on the Gjs exposed API.

Wouldn't the Gjs APIs be better than subprocess? I would expect them to be more stable than the text output of `wpctl` but I'm not familiar with Gnome programming.

I played with the Gjs Wireplumber APIs a little bit without much success. But I think there might be some sample Lua code that might be able to serve as a reference.

**Comment by danigm** [comment]
> Wouldn't the Gjs APIs be better than subprocess? I would expect them to be more stable than the text output of wpctl but I'm not familiar with Gnome programming.

@bmaupin yes, of course, it could be better with Gjs, but if it's too hard, maybe we can workaround with the subprocess idea.
